File: 041-060/challenge_053.py
'''
Implement a queue using two stacks. Recall that a queue is a FIFO (first-in, first-out) data structure with the following methods:
enqueue, which inserts an element into the queue, and dequeue, which removes it.
'''

class Stack:
    class StackNode:
        def __init__(self, data):
            self.data = data
            self.next = None

    def __init__(self):
        self.top = None

    def push(self, data):
        new_node = self.StackNode(data)
        new_node.next = self.top
        self.top = new_node
        return self

    def pop(self):
        if self.top == None:
            return None
        
        pop_node = self.top
        self.top = self.top.next
        return pop_node.data

class Queue:
    # hold_stack is the stack that contains the queue. flip_stack is empty except when dequeue is called.
    def __init__(self):
        self.hold_stack = Stack()
        self.flip_stack = Stack()
        self._size = 0

    def __len__(self):
        return self._size

    def enqueue(self, data):
        self.hold_stack.push(data)
        self._size += 1
        return self

    def dequeue(self):
        if self._size <= 0:
            return None
        # Move everything to the flip_stack, remove the bottom from hold_stack, and flip back.
        for i in range(0, self._size-1):
            self.flip_stack.push(self.hold_stack.pop())
        dequeue_data = self.hold_stack.pop()
        for i in range(0, self._size-1):
            self.hold_stack.push(self.flip_stack.pop())
        self._size -= 1
        return dequeue_data

if __name__ == "__main__":
    q = Queue()
    q.enqueue(1).enqueue(2).enqueue(3).enqueue(4)
    print(len(q))
    print(q.dequeue())
    print(len(q))

File: js/041-060/challenge_053.js
/*
Implement a queue using two stacks. Recall that a queue is a FIFO (first-in, first-out) data structure with the following methods:
enqueue, which inserts an element into the queue, and dequeue, which removes it.
*/

class Stack {
    #top = null;

    constructor() {
    }

    // Using an object structure that contains the next stack node as the stack node.
    push(data) {
        const lastTop = this.#top;
        this.#top = {
            'data' : data,
            'next' : lastTop
        };
        return;
    }

    pop() {
        if (this.#top == null) {
            return null;
        }

        let outData = this.#top.data;
        this.#top = this.#top.next;

        return outData;
    }

    isEmpty() {
        return this.#top == null;
    }

    // Print Methods
    toString() {
        let stackStr = "";
        let nodePtr = this.#top;

        while (nodePtr != null) {
            stackStr = stackStr.concat(nodePtr.data.toString());
            if (nodePtr.next != null) {
                stackStr = stackStr.concat("->");
            }
            nodePtr = nodePtr.next;
        }

        return stackStr;
    }

    printStack() {
        console.log(this.toString().concat("<Stack Top>"));
        return;
    }
}

class Queue {
    #holdStack = new Stack();
    #flipStack = new Stack();

    constructor() {
    }

    // Enqueue pushes the main holdStack to the flipStack from top to bottom, pushes the data to the holdStack, then flips the flipStack back.
    enqueue(data) {
        while (!this.#holdStack.isEmpty()) {
            this.#flipStack.push(this.#holdStack.pop());
        }

        this.#holdStack.push(data);

        while(!this.#flipStack.isEmpty()) {
            this.#holdStack.push(this.#flipStack.pop());
        }

        return;
    }

    // Dequeue simply calls pop() on the holdStack.
    dequeue() {
        return this.#holdStack.pop();
    }

    isEmpty() {
        return this.#holdStack.isEmpty();
    }

    // Print Methods
    toString() {
        return this.#holdStack.toString();
    }

    printQueue() {
        console.log("<Queue Start>".concat(this.#holdStack.toString()));
        return;
    }
}

const q = new Queue();
q.enqueue(1);
q.enqueue(2);
q.enqueue(3);
q.printQueue();
q.enqueue(4);
q.printQueue();
q.dequeue();
q.printQueue();